auth0/provider_test.go

```diff
@@ -71,3 +73,92 @@ func TestProvider_debugDefaults(t *testing.T) {
 		}
 	}
 }
+
+func TestProvider_configValidation(t *testing.T) {
+	testCases := []struct {
+		name           string
+		environment    map[string]string
+		resourceConfig map[string]interface{}
+		expectedErrors []error
+	}{
+		{
+			name:           "missing client id",
+			environment:    map[string]string{"AUTH0_DOMAIN": "test", "AUTH0_CLIENT_SECRET": "test"},
+			expectedErrors: []error{errors.New("\"client_secret\": all of `client_id,client_secret` must be specified")},
+		},
+		{
+			name:           "missing client secret",
+			environment:    map[string]string{"AUTH0_DOMAIN": "test", "AUTH0_CLIENT_ID": "test"},
+			expectedErrors: []error{errors.New("\"client_id\": all of `client_id,client_secret` must be specified")},
+		},
+		{
+			name:           "conflicting auth0 client and management token without domain",
+			resourceConfig: map[string]interface{}{"client_id": "test", "client_secret": "test", "api_token": "test"},
+			environment:    map[string]string{},
+			expectedErrors: []error{
+				errors.New("\"domain\": required field is not set"),
+				errors.New("\"client_id\": conflicts with api_token"),
+				errors.New("\"client_secret\": conflicts with api_token"),
+				errors.New("\"api_token\": conflicts with client_id"),
+			},
+		},
+		{
+			name:           "valid auth0 client",
+			resourceConfig: map[string]interface{}{"domain": "valid_domain", "client_id": "test", "client_secret": "test"},
+			environment:    map[string]string{},
+			expectedErrors: nil,
+		},
+		{
+			name:           "valid auth0 token",
+			resourceConfig: map[string]interface{}{"domain": "valid_domain", "api_token": "test"},
+			environment:    map[string]string{},
+			expectedErrors: nil,
+		},
+		{
+			name:           "valid auth0 token from environment",
+			resourceConfig: map[string]interface{}{"domain": "valid_domain"},
+			environment:    map[string]string{"AUTH0_API_TOKEN": "test"},
+			expectedErrors: nil,
+		},
+	}
+
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			for k, v := range test.environment {
+				os.Unsetenv(k)
+				os.Setenv(k, v)
+			}
```

Comment:
I ran the full test suite locally and unfortunately having both resourceConfig set up and the env vars messes up with running everything in 1 go as the env vars will be unset for the subsequent acceptance tests. To fix this we could either:

A) keep track of env vars before setting the ones in the tests and replace the original values after they run
B) test only with the resourceConfig values 

Apologies for having missed this before @kevinschoonover. Would you be able to remove the env vars from these tests and just keep the resourceConfig as that should already cover our intended behavior change in the auth mechanism?

Reply:
I think we need to do A) because I ran the acceptance test again removing this code and its pulling client_id and client_secret from the environment breaking the existing tests.

```bash
> AUTH0_DOMAIN="https://kschoon.auth0.com" AUTH0_CLIENT_ID="<id>" AUTH0_CLIENT_SECRET="<id>" make testacc
=== RUN   TestProvider_configValidation
=== RUN   TestProvider_configValidation/missing_client_id
    provider_test.go:134: actual did not match expected. len(expected) != len(actual). expected: ["client_secret": all of `client_id,client_secret` must be specified], actual: ["client_secret": all of `client_id,client_secret` must be specified "client_id": all of `client_id,client_secret` must be specified]
=== RUN   TestProvider_configValidation/missing_client_secret
    provider_test.go:134: actual did not match expected. len(expected) != len(actual). expected: ["client_id": all of `client_id,client_secret` must be specified], actual: ["client_id": all of `client_id,client_secret` must be specified "client_secret": all of `client_id,client_secret` must be specified]
=== RUN   TestProvider_configValidation/conflicting_auth0_client_and_management_token_without_domain
    provider_test.go:134: actual did not match expected. len(expected) != len(actual). expected: ["domain": required field is not set "client_id": conflicts with api_token "client_secret": conflicts with api_token "api_token": conflicts with client_id], actual: ["api_token": conflicts with client_id "client_id": conflicts with api_token "client_secret": conflicts with api_token]
=== RUN   TestProvider_configValidation/valid_auth0_client
=== RUN   TestProvider_configValidation/valid_auth0_token
    provider_test.go:134: actual did not match expected. len(expected) != len(actual). expected: [], actual: ["client_secret": all of `client_id,client_secret` must be specified "client_id": all of `client_id,client_secret` must be specified]
--- FAIL: TestProvider_configValidation (0.06s)
    --- FAIL: TestProvider_configValidation/missing_client_id (0.01s)
    --- FAIL: TestProvider_configValidation/missing_client_secret (0.01s)
    --- FAIL: TestProvider_configValidation/conflicting_auth0_client_and_management_token_without_domain (0.01s)
    --- PASS: TestProvider_configValidation/valid_auth0_client (0.01s)
    --- FAIL: TestProvider_configValidation/valid_auth0_token (0.01s)
```